Fixes #241: Added support for Neo4j Vector Index

vga91 opened the pull request:
Fixes #241: Added support for Neo4j Vector Index

This commit brings support for Neo4j graph database in general, and uses the vector index functionality, generally available since version 5.13.

Mostly aligned with the existing WeaviateEmbeddingStoreImpl implementation and tests.
The tests have some additional Neo4j node assertion to check that the nodes involved are correctly created.

The module creates indexes, i.e. `"CALL db.index.vector.createNodeIndex(<indexName>, <label>, <embeddingProperty>, <dimension>, <distanceType>)"`, if needed, for the vector search .

The required configurations are:
- the Neo4j index dimension parameter
- the Neo4j Java Driver connection instance
- as an alternative to the Neo4j Java Driver, we can create a `Neo4jEmbeddingStore.builder().withBasicAuth(<url>, <username>, <password>)`, which will create a Driver connection instance under the hood

It is possible to customize, via the builder:
- the index name (with default `langchain-embedding-index`)
- the Neo4j node label  (with default `Document`)
- the Neo4j property key which save the embeddings  (with default `embeddingProp`)
- the Neo4j index distanceType parameter
- the metadata prefix (with default `metadata.`)
- the text property key (with default `text`), which store the text field of the `TextSegment.java`


Created an example PR as well, on `langchain4j-examples` repo: https://github.com/langchain4j/langchain4j-examples/pull/23

vga91 committed:
Fixes #241: Added support for Neo4j Vector Index

jexp commented on langchain4j-neo4j/src/main/java/dev/langchain4j/store/embedding/neo4j/Neo4jEmbeddingUtils.java:
```diff
@@ -0,0 +1,82 @@
+    public static final String ID_ROW_KEY = "id";
+    
+    /* default configs */
+    public static final String DEFAULT_EMBEDDING_PROP = "embeddingProp";
```
change to `embedding`

vga91 replied:
Changed

jexp commented on langchain4j-neo4j/src/main/java/dev/langchain4j/store/embedding/neo4j/Neo4jEmbeddingUtils.java:
```diff
@@ -0,0 +1,82 @@
+            TextSegment segment = embedded.get(idx);
+            properties.put(store.getText(), segment.text());
+            Map<String, String> metadata = segment.metadata().asMap();
+            metadata.forEach((k, v) -> properties.put(store.getMetadataPrefix() + k, Values.value(v)));
```
can we have the default option to be the metadata-keys without a prefix, and only if a prefix is configured it is used?

vga91 replied:
Now the default prefix is `""`

jexp commented on langchain4j-neo4j/src/main/java/dev/langchain4j/store/embedding/neo4j/Neo4jEmbeddingUtils.java:
```diff
@@ -0,0 +1,82 @@
+            metadata.forEach((k, v) -> properties.put(store.getMetadataPrefix() + k, Values.value(v)));
+        }
+
+        row.put(EMBEDDINGS_ROW_KEY, Values.value(embedding.vector()));
```
should we use a float[] here, need to check with the drivers team.

vga91 committed:
changed default meta prefix and embedding prop

jexp commented on langchain4j-neo4j/src/main/java/dev/langchain4j/store/embedding/neo4j/Neo4jEmbeddingStore.java:
```diff
@@ -0,0 +1,244 @@
+     * starting from uri, user and password
+     */
+    public static class Neo4jEmbeddingStoreBuilder {
+        public Neo4jEmbeddingStoreBuilder withBasicAuth(String uri, String user, String password) {
```
Should also allow for database name

vga91 replied:
Added databaseName config as well

jexp commented on langchain4j-neo4j/src/main/java/dev/langchain4j/store/embedding/neo4j/Neo4jEmbeddingStore.java:
```diff
@@ -0,0 +1,244 @@
+     */
+    @Builder
+    public Neo4jEmbeddingStore(
+            SessionConfig config,
```
Good question if we want to leak the session config from neo4j into the client code. I would probably start with database name and then add a builder methods for other options.

Should the builder constructor be public?

vga91 replied:
Added databaseName config as well

jexp commented on langchain4j-neo4j/src/main/java/dev/langchain4j/store/embedding/neo4j/Neo4jEmbeddingStore.java:
```diff
@@ -0,0 +1,244 @@
+        
+        /* required configs */
+        ensureNotNull(driver, "driver");
+        ensureNotNull(dimension, "dimension");
```
should also check that dimension is > 0 and  <= 4096 (currently Neo4j only supports sizes to 2048) but will increase that.

jexp commented on langchain4j-neo4j/src/main/java/dev/langchain4j/store/embedding/neo4j/Neo4jEmbeddingStore.java:
```diff
@@ -0,0 +1,244 @@
+        this.dimension = dimension;
+        
+        /* optional configs */
+        this.config = getOrDefault(config, SessionConfig.forDatabase("neo4j"));
```
make database configurable

jexp commented on langchain4j-neo4j/src/main/java/dev/langchain4j/store/embedding/neo4j/Neo4jEmbeddingStore.java:
```diff
@@ -0,0 +1,244 @@
+    }
+
+    @Override
+    public List<EmbeddingMatch<TextSegment>> findRelevant(Embedding referenceEmbedding, int maxResults, double minScore) {
```
what is the default for minScore?

jexp commented on langchain4j-neo4j/src/main/java/dev/langchain4j/store/embedding/neo4j/Neo4jEmbeddingStore.java:
```diff
@@ -0,0 +1,244 @@
+        try (var session = session()) {
+            String statement = """
+                            UNWIND $rows AS row
+                            MERGE (u:%1$s {%2$s: row.%2$s})
```
add backticks around label and property name, and row values.
and sanitize label and property names to avoid injection attacks.

vga91 replied:
I added `sanitize()` to labels and properties.
I didn't add backticks here, since together with `sanitize`, it could create double-backticks

jexp commented on langchain4j-neo4j/src/main/java/dev/langchain4j/store/embedding/neo4j/Neo4jEmbeddingStore.java:
```diff
@@ -0,0 +1,244 @@
+        try (var session = session()) {
+            String statement = """
+                            UNWIND $rows AS row
+                            MERGE (u:%1$s {%2$s: row.%2$s})
```
if we merge on the id - property we should make sure that a constraint exists for that id property and create if it doesn't.

vga91 replied:
Added a [`CREATE CONSTRAINT` statement](https://github.com/langchain4j/langchain4j/pull/282/commits/36cd2f46c1280e53068bd2b9cd26941e17dfb0ac#diff-89b713731c6e14e29b6208dc8c14a92aff1a1baad94b44b8a9b388db8be97eb9R249)

jexp commented on langchain4j-neo4j/src/main/java/dev/langchain4j/store/embedding/neo4j/Neo4jEmbeddingStore.java:
```diff
@@ -0,0 +1,244 @@
+                            SET u += row.%3$s
+                            WITH row, u
+                            CALL db.create.setNodeVectorProperty(u, $embeddingProperty, row.%4$s)
+                            RETURN *""".formatted(
```
we should return count(*) to avoid huge results coming back (think of all the input data and all the nodes with their properties coming back).

jexp commented on langchain4j-neo4j/src/main/java/dev/langchain4j/store/embedding/neo4j/Neo4jEmbeddingStore.java:
```diff
@@ -0,0 +1,244 @@
+                    "embeddingProperty", this.embeddingProperty
+            );
+            
+            session.run(statement, params).consume();
```
we also need to batch this from the outside, so that we limit each tx to 10k rows or so. and use 
tx.executeWrite() for each batch.

jexp commented on langchain4j-neo4j/src/main/java/dev/langchain4j/store/embedding/neo4j/Neo4jEmbeddingStore.java:
```diff
@@ -0,0 +1,244 @@
+    private boolean indexExists() {
+        try (var session = session()) {
+            Map<String, Object> params = Map.of("name", this.indexName);
+            return session.run("SHOW INDEX WHERE type = 'VECTOR' AND name = $name", params)
```
also verify that it has the same label + embedding property

jexp commented on langchain4j-neo4j/src/main/java/dev/langchain4j/store/embedding/neo4j/Neo4jEmbeddingStore.java:
```diff
@@ -0,0 +1,244 @@
+            session.run("CALL db.index.vector.createNodeIndex($indexName, $label, $embeddingProperty, $dimension, $distanceType)",
+                    params);
+
+            session.run("CALL db.awaitIndexes()").consume();
```
the default is 300s (5min) here, not sure if we want to pass in an explicit value that is large enough for the index population to finish in case there is already data in the db.

vga91 replied:
Changed to 60

jexp commented on langchain4j-neo4j/src/main/java/dev/langchain4j/store/embedding/neo4j/Neo4jEmbeddingStore.java:
```diff
@@ -0,0 +1,244 @@
+ * Annotated with `@Getter` to be used in {@link Neo4jEmbeddingUtils}
+ */
+@Getter
+public class Neo4jEmbeddingStore implements EmbeddingStore<TextSegment> {
```
is there any API in langchain4j to allow existing data in the database to be embedded and indexed?
what does the whole roundtrip look like in an example?

jexp commented on langchain4j-neo4j/src/main/java/dev/langchain4j/store/embedding/neo4j/Neo4jEmbeddingStore.java:
```diff
@@ -0,0 +1,244 @@
+                            CALL db.create.setNodeVectorProperty(u, $embeddingProperty, row.%4$s)
+                            RETURN *""".formatted(
+                    this.label,
+                    ID_ROW_KEY,
```
should be configurable

jexp commented on langchain4j-neo4j/src/main/java/dev/langchain4j/store/embedding/neo4j/Neo4jEmbeddingStore.java:
```diff
@@ -0,0 +1,244 @@
+        
+        /* optional configs */
+        this.config = getOrDefault(config, SessionConfig.forDatabase("neo4j"));
+        this.label = getOrDefault(label, DEFAULT_LABEL);
```
id should also be configurable

vga91 replied:
Added

jexp commented on langchain4j-neo4j/src/main/java/dev/langchain4j/store/embedding/neo4j/Neo4jEmbeddingStore.java:
```diff
@@ -0,0 +1,244 @@
+        /* optional configs */
+        this.config = getOrDefault(config, SessionConfig.forDatabase("neo4j"));
+        this.label = getOrDefault(label, DEFAULT_LABEL);
+        this.embeddingProperty = getOrDefault(property, DEFAULT_EMBEDDING_PROP);
```
I'm missing the builder methods for all these other config options? are they somewhere else?

vga91 replied:
The getter and builder method are auto-generate via[Lombok annotations.](https://github.com/langchain4j/langchain4j/pull/282/commits/36cd2f46c1280e53068bd2b9cd26941e17dfb0ac#diff-89b713731c6e14e29b6208dc8c14a92aff1a1baad94b44b8a9b388db8be97eb9R86)

jexp commented on langchain4j-neo4j/src/main/java/dev/langchain4j/store/embedding/neo4j/Neo4jEmbeddingUtils.java:
```diff
@@ -0,0 +1,83 @@
+    
+    /* not-configurable strings, used in `UNWIND $rows ...` statement */
+    public static final String EMBEDDINGS_ROW_KEY = "embeddingRow";
+    public static final String ID_ROW_KEY = "id";
```
should be also configurable

jexp commented on langchain4j-neo4j/src/main/java/dev/langchain4j/store/embedding/neo4j/Neo4jEmbeddingUtils.java:
```diff
@@ -0,0 +1,83 @@
+    /* default configs */
+    public static final String DEFAULT_EMBEDDING_PROP = "embedding";
+    public static final String PROPS = "props";
+    public static final String DEFAULT_IDX_NAME = "langchain-embedding-index";
```
probably use a similar namings like we do in the langchain integration?

jexp commented on langchain4j-neo4j/src/main/java/dev/langchain4j/store/embedding/neo4j/Neo4jEmbeddingStore.java:
```diff
@@ -0,0 +1,244 @@
+						CALL db.index.vector.queryNodes($indexName, $maxResults, $embeddingValue)
+						YIELD node, score
+						WHERE score >= $minScore
+						RETURN node, score
```
can we make configurable which properties we want to return or the full node?
because if we return all the huge properties like the embedding + big texts that can become a lot

we should also add the option like in langchain integration to provide a retrievalQuery in the config that is applied after the node/score is retrieved from the index so that the user can optionally find more contextually relevant data in the neighbourhood.

vga91 replied:
Added a retrievalQuery config, with default:
```
RETURN properties(node) AS metadata, node.id AS id, node.text AS text, node.embedding AS embedding, score
```

jexp commented on langchain4j-neo4j/src/main/java/dev/langchain4j/store/embedding/neo4j/Neo4jEmbeddingUtils.java:
```diff
@@ -0,0 +1,83 @@
+    
+
+    public static EmbeddingMatch<TextSegment> toEmbeddingMatch(Neo4jEmbeddingStore store, Record neo4jRecord) {
+        var node = neo4jRecord.get("node").asNode();
```
we should make configurable what we actually want to return from the database and then also from the find method, otherwise it might be a lot of data that is returned.

jexp commented on langchain4j-neo4j/src/main/java/dev/langchain4j/store/embedding/neo4j/Neo4jEmbeddingUtils.java:
```diff
@@ -0,0 +1,83 @@
+        node.keys().forEach(key -> {
+            Set<String> notMetaKeys = Set.of(ID_ROW_KEY, store.getEmbeddingProperty(), store.getText());
+            if (!notMetaKeys.contains(key)) {
+                metaData.put(key.replace(store.getMetadataPrefix(), ""), node.get(key).asString());
```
asString or asObject?

jexp commented on langchain4j-neo4j/src/test/java/dev/langchain4j/store/embedding/neo4j/Neo4jEmbeddingStoreTest.java:
```diff
@@ -0,0 +1,371 @@
+
+    @AfterEach
+    void afterEach() {
+        session.run("MATCH (n) DETACH DELETE n");
```
also drop the index?

vga91 replied:
Added

jexp commented on langchain4j-neo4j/src/test/java/dev/langchain4j/store/embedding/neo4j/Neo4jEmbeddingStoreTest.java:
```diff
@@ -0,0 +1,371 @@
+    void should_add_embedding_with_id() {
+
+        String id = randomUUID();
+        Embedding embedding = embeddingModel.embed(randomUUID()).content();
```
different id ??

jexp replied:
oh this is meant to be a text?

jexp commented on langchain4j-neo4j/src/test/java/dev/langchain4j/store/embedding/neo4j/Neo4jEmbeddingStoreTest.java:
```diff
@@ -0,0 +1,371 @@
+
+    @Test
+    void should_add_embedding() {
+        Embedding embedding = embeddingModel.embed(randomUUID()).content();
```
perhaps use something else as text ? using randomUUID() in two places is confusing :)

jexp reviewed (changes requested):
Looks good, thanks so much @vga91 
Please have a look at my comments.

The configuration of what is returned / retrievalQuery and some other extra things can wait for a later PR.

Id configuration (probably the id-constraint), backticks etc. would be good to have in the first PR for security/performance reasons.

I didn't see any docs or an end-to-end example. Not sure how that works in langchain4j.

jexp commented on langchain4j-neo4j/pom.xml:
```diff
@@ -0,0 +1,93 @@
+
+    <properties>
+        <testcontainers.version>1.19.0</testcontainers.version>
+        <maven.compiler.source>17</maven.compiler.source>
```
Not for the current versions that have vector support, those are on 17.
Also this uses the client drivers, so it's independent of the server installation.

jexp replied:
Which is at 17 too
https://neo4j.com/docs/java-manual/current/get-started/#java-driver-get-started-installation

vga91 committed:
Various changes review

vga91 commented on langchain4j-neo4j/src/main/java/dev/langchain4j/store/embedding/neo4j/Neo4jEmbeddingStore.java:
```diff
@@ -0,0 +1,244 @@
+        ensureNotNull(driver, "driver");
+        ensureNotNull(dimension, "dimension");
+
+        this.driver = driver;
```
Changed

vga91 commented:
Sorry for the delay. 
I just pushed the following changes:
- databaseName config
- `"id"` property config
- added auto-constraint for the property `"id"`
- dimension: check that is > 0 and <= 4096 
- added sanitize to label and property name
- added batch every 10,000 rows
- added check to `SHOW INDEX WHERE type = 'VECTOR' ..`
- changed `CALL db.awaitIndexes()` to `CALL db.awaitIndexes(60)`
- added `retrievalQuery` to customize what is returned in `CALL db.index.vector.queryNodes`

jexp commented on langchain4j-neo4j/src/main/java/dev/langchain4j/store/embedding/neo4j/Neo4jEmbeddingStore.java:
```diff
@@ -0,0 +1,310 @@
+    }
+
+    private void bulk(List<String> ids, List<Embedding> embeddings, List<TextSegment> embedded) {
+        Collection<List<Map<String, Object>>> rowsBatched = getRowsBatched(this, ids, embeddings, embedded);
```
Would it be better to return a stream here instead of a collection? So that we stream one chunk at a time? And don't allocate all the memory upfront?

jexp commented on langchain4j-neo4j/src/main/java/dev/langchain4j/store/embedding/neo4j/Neo4jEmbeddingUtils.java:
```diff
@@ -0,0 +1,113 @@
+        return row;
+    }
+
+    public static Collection<List<Map<String, Object>>> getRowsBatched(Neo4jEmbeddingStore store, List<String> ids, List<Embedding> embeddings, List<TextSegment> embedded) {
```
perhaps change to return a stream of chunks, where don't use a grouping collector but a IntStream of startIndex*batchSize + batchSize to create a chunk at a time.

jexp reviewed (approved):
Looks good to me.

Only 2 small things

- can you change the chunking to streaming
- add some unit tests for Neo4jEmbeddingUtils (for the chunking esp. off by sizes) and the sanitation

vga91 committed:
changed batching logic - added utils tests

jexp reviewed (approved):
Looks good, thanks so much @vga91 
Good to go for my @langchain4j

jexp commented:
@langchain4j not sure how to resolve all the conversations, there is no "resolve all" and the individual ones don't show "Resolve" Menu/Button for me.

jexp reviewed (approved):
Good to go for me @langchain4j (but the "review required" doesn't go away)

dliubarskyi committed:
Merge branch 'main' into neo4j-vector-index-integration

dliubarskyi committed:
Merge branch 'main' into neo4j-vector-index-integration

Commit:
Revert "excluded neo4j module from java 8 and 11 builds"

This reverts commit f5039b6eea7d784850a6080c073c97d24b06a89e.

Commit:
temporarily adding missing util method

vga91 commented on langchain4j-neo4j/src/main/java/dev/langchain4j/store/embedding/neo4j/Neo4jDistanceType.java:
```diff
@@ -0,0 +1,16 @@
+    COSINE("cosine"),
+    
+    /* euclidean distance */
+    EUCLIDEAN("euclidean");
```
Do you mean, removing the Neo4jDistanceType.java class at all and expliciting "cosine" everywhere, 
or just removing `EUCLIDEAN("euclidean");` from here?

vga91 replied:
removed

vga91 commented on langchain4j-neo4j/src/main/java/dev/langchain4j/store/embedding/neo4j/Neo4jEmbeddingStore.java:
```diff
@@ -0,0 +1,311 @@
+        Stream<List<Map<String, Object>>> rowsBatched = getRowsBatched(this, ids, embeddings, embedded);
+
+        rowsBatched.forEach(rows -> {
+            try (var session = session()) {
```
Yes actually it's better, I'll change it

vga91 commented on langchain4j-neo4j/src/main/java/dev/langchain4j/store/embedding/neo4j/Neo4jEmbeddingStore.java:
```diff
@@ -0,0 +1,311 @@
+                        "embeddingProperty", this.embeddingProperty
+                );
+
+                session.executeWrite(tx -> tx.run(statement, params).consume());
```
Yes the intention is to execute a transaction for each batch

vga91 committed:
changes 2nd review

vga91 commented on langchain4j-neo4j/src/main/java/dev/langchain4j/store/embedding/neo4j/Neo4jEmbeddingUtils.java:
```diff
@@ -0,0 +1,120 @@
+                    String invalidSanitizeValue = String.format("The value %s, to assign to configuration %s, cannot be safely quoted",
+                            value,
+                            config);
+                    throw new RuntimeException(invalidSanitizeValue);
```
changed

vga91 commented on langchain4j-neo4j/src/main/java/dev/langchain4j/store/embedding/neo4j/Neo4jEmbeddingStore.java:
```diff
@@ -0,0 +1,311 @@
+            session.run("CALL db.index.vector.createNodeIndex($indexName, $label, $embeddingProperty, $dimension, $distanceType)",
+                    params);
+
+            session.run("CALL db.awaitIndexes(60)").consume();
```
60 is a quite high value, but indeed a configuration value is better.
Added an `awaitIndexTimeout` config

vga91 commented:
@langchain4j 

I've made the various changes, except the EUCLIDEAN stuff I asked about.

I'm not sure why the build fails, executing a `mvn --batch-mode install -DskipITs` locally, with temurin `21.0.1-tem` (which should be the same version) it works, and the lombok version should be compatible with JDK 21..

It's probably due to [the version](https://github.com/langchain4j/langchain4j/pull/282/files#diff-a88727153e9e65be269b619a6a6351f881a574c0aec035bcbcb90d5c97fb2e14R10), which was 0.23 and now I changed it to 0.24?

Anyway from what I see, the CI doesn't start unless all the conversations are resolved, so I guess we could check it out after doing that

dliubarskyi committed:
Merge branch 'main' into neo4j-vector-index-integration

vga91 committed:
removed Neo4jDistanceType and assumed indexType = 'cosine'

vga91 committed:
added exclusion to java_version 8 and 11

vga91 commented:
@langchain4j I created a PR that should solve the problem: https://github.com/langchain4j/langchain4j/pull/368

vga91 commented:
@langchain4j I created another PR for it https://github.com/langchain4j/langchain4j/pull/441
